File: p3/three.hs

module Three where

main = print . max_factor $ 600851475143

-- Factors out fac from num.
-- If fac is the only factor of num, returns num.
remove :: Integral i => i -> i -> i
remove num fac = if num > fac && num `mod` fac == 0 then
                     remove (num `div` fac) fac
                 else num

-- Integer square root function, rounding up.
isqrt :: Integral i => i -> i
isqrt n = ceiling . sqrt . fromIntegral $ n

-- Finds the largest prime factor of n.
max_factor :: Integral i => i -> i
max_factor n = foldl remove n (2:[3,5..isqrt n])

File: p3/three.py
#!/usr/bin/env python3

# What is the largest prime factor of the number 600851475143 ?

from math import sqrt

def max_factor(num):
    i = 2
    root = int(sqrt(num))
    while i <= root:
        if num == i:
            break
        elif num % i == 0:
            num = num // i
        else:
            i += 1
    return num

if __name__ == '__main__':
    print(max_factor(600851475143))
